python/cudf/cudf/core/join/_join_helpers.py

```diff
@@ -0,0 +1,206 @@
+    lcol: ColumnBase, rcol: ColumnBase, how: str
+) -> Tuple[ColumnBase, ColumnBase]:
+    # cast the keys lcol and rcol to a common dtype
+    # when at least one of them is a categorical type
+    ltype, rtype = lcol.dtype, rcol.dtype
+
+    if isinstance(lcol, cudf.core.column.CategoricalColumn) and isinstance(
+        rcol, cudf.core.column.CategoricalColumn
+    ):
+        # if both are categoricals, logic is complicated:
+        return _match_categorical_dtypes_both(lcol, rcol, how)
+
+    if isinstance(ltype, CategoricalDtype):
+        if how in {"left", "leftsemi", "leftanti"}:
+            return lcol, rcol.astype(ltype)
+        common_type = ltype.categories.dtype
+    elif isinstance(rtype, CategoricalDtype):
+        common_type = rtype.categories.dtype
+    return lcol.astype(common_type), rcol.astype(common_type)
+
+
+def _match_categorical_dtypes_both(
+    lcol: CategoricalColumn, rcol: CategoricalColumn, how: str
+) -> Tuple[ColumnBase, ColumnBase]:
+    # The commontype depends on both `how` and the specifics of the
+    # categorical variables to be merged.
+
+    ltype, rtype = lcol.dtype, rcol.dtype
+
+    # when both are ordered and both have the same categories,
+    # no casting required:
+    if ltype == rtype:
+        return lcol, rcol
+
+    # Merging categorical variables when only one side is ordered is
+    # ambiguous and not allowed.
+    if ltype.ordered != rtype.ordered:
+        raise TypeError(
+            "Merging on categorical variables with mismatched"
+            " ordering is ambiguous"
+        )
+
+    if ltype.ordered and rtype.ordered:
+        # if we get to here, categories must be what causes the
+        # dtype equality check to fail. And we can never merge
+        # two ordered categoricals with different categories
+        raise TypeError(
+            f"{how} merge between categoricals with "
+            "different categories is only valid when "
+            "neither side is ordered"
+        )
+
+    # the following should now always hold
+    assert not ltype.ordered and not rtype.ordered
+
+    if how == "inner":
+        # cast to category types -- we must cast them back later
+        return _match_join_keys(
+            lcol.cat()._decategorize(), rcol.cat()._decategorize(), how,
+        )
+    elif how in {"left", "leftanti", "leftsemi"}:
+        # always cast to left type
+        return lcol, rcol.astype(ltype)
+    else:
+        # merge categories
+        merged_categories = cudf.concat(
+            [ltype.categories, rtype.categories]
+        ).unique()
+        common_type = cudf.CategoricalDtype(
+            categories=merged_categories, ordered=False
+        )
+        return lcol.astype(common_type), rcol.astype(common_type)
+
+
+def _coerce_to_tuple(obj):
+    if hasattr(obj, "__iter__") and not isinstance(obj, str):
+        return tuple(obj)
+    else:
+        return (obj,)
+
+
+def _coerce_to_list(obj):
+    return list(_coerce_to_tuple(obj))
```

Comment:
Is this performance sensitive where we'd care about the overhead of tuple --> list conversion?

Reply:
Took @vyasr's advice and just got rid of `_coerce_to_list`. There was only one place we needed it and now we do `list(_coerce_to_tuple(...))` explicitly there. I don't think it's super perf sensitive. Certainly hasn't shown up in my profiles.